src/block/Hopper.php

```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered()){
+			// If a hopper is powered it is deactivated and won't push, pull or pick up items.
+			return;
+		}
+
+		if($transferCooldown > 0){
+			// Skipping this tick because the hopper is still on cooldown of 8 ticks.
+			return;
+		}
+
+		$success = $this->push($tile);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		if($origin instanceof Container){
+			$success |= $this->pull($tile, $origin);
+		}else{
+			$success |= $this->pickup($tile);
+		}
+		// The cooldown is only set back to 8 ticks if the hopper has done anything.
+		if((bool) $success){
+			$tile->setTransferCooldown(8);
```

Comment:
This should be moved to a constant. Hardcoding the value is not good practice.

Reply:
Hm
I am unsure about a clear name for that constant.
TRANSFER_COOLDOWN doesn't suit in my opinion, because the variable that stores the current cooldown of the hopper already has just a slightly different name which could confuse.
But I don't think that MAX_TRANSFER_COOLDOWN is so perfect either.